netadapter/router.go

```diff
@@ -5,27 +5,83 @@ import (
 	"github.com/pkg/errors"
 )
 
+// OnIDReceivedHandler is a function that is to be called
+// once a new Connection is sends us its ID.
+type OnIDReceivedHandler func(id *ID)
+
 // Router routes messages by type to their respective
 // input channels
 type Router struct {
-	routes map[string]chan<- wire.Message
+	inputRoutes         map[string]chan wire.Message
+	outputRoute         chan wire.Message
+	onIDReceivedHandler OnIDReceivedHandler
+}
+
+// NewRouter creates a new empty router
+func NewRouter() *Router {
+	return &Router{
+		inputRoutes: make(map[string]chan wire.Message),
+		outputRoute: make(chan wire.Message),
+	}
+}
+
+// SetOnIDReceivedHandler sets the onIDReceivedHandler function for
+// this router
+func (r *Router) SetOnIDReceivedHandler(onIDReceivedHandler OnIDReceivedHandler) {
+	r.onIDReceivedHandler = onIDReceivedHandler
 }
 
 // AddRoute registers the messages of types `messageTypes` to
-// be routed to the given `inChannel`
-func (r *Router) AddRoute(messageTypes []string, inChannel chan<- wire.Message) error {
+// be routed to the given `inputChannel`
+func (r *Router) AddRoute(messageTypes []string, inputChannel chan wire.Message) error {
 	for _, messageType := range messageTypes {
-		if _, ok := r.routes[messageType]; ok {
+		if _, ok := r.inputRoutes[messageType]; ok {
 			return errors.Errorf("a route for '%s' already exists", messageType)
 		}
-		r.routes[messageType] = inChannel
+		r.inputRoutes[messageType] = inputChannel
+	}
+	return nil
+}
+
+// RemoveRoute unregisters the messages of types `messageTypes` from
+// the router
+func (r *Router) RemoveRoute(messageTypes []string) error {
+	for _, messageType := range messageTypes {
+		if _, ok := r.inputRoutes[messageType]; !ok {
+			return errors.Errorf("a route for '%s' does not exist", messageType)
+		}
+		delete(r.inputRoutes, messageType)
 	}
 	return nil
 }
 
-// RouteMessage sends the given message to the correct input
+// RouteInputMessage sends the given message to the correct input
 // channel as registered with AddRoute
-func (r *Router) RouteMessage(message wire.Message) {
-	routeInChannel := r.routes[message.Command()]
+func (r *Router) RouteInputMessage(message wire.Message) {
+	routeInChannel := r.inputRoutes[message.Command()]
 	routeInChannel <- message
 }
+
+// TakeOutputMessage takes the next output message from
+// the output channel
+func (r *Router) TakeOutputMessage() wire.Message {
+	return <-r.outputRoute
+}
+
+// RegisterID registers the remote connection's ID
+func (r *Router) RegisterID(id *ID) {
+	r.onIDReceivedHandler(id)
+}
+
+// Close shuts down the router by closing all registered
+// input channels
+func (r *Router) Close() error {
+	inChannels := make(map[chan<- wire.Message]struct{})
+	for _, inChannel := range r.inputRoutes {
+		inChannels[inChannel] = struct{}{}
+	}
+	for inChannel := range inChannels {
```

Comment:
Why the double loop?

Reply:
Routes generally point N messages to 1 channel and I don't want to call close() on a channel more than once.